README.md
# Dart wrapper for Keycloak JavaScript Adapter

Provide a more Dart-ish interface for [Keycloak JS Adapter](https://www.keycloak.org/docs/latest/securing_apps/index.html#_javascript_adapter).

## Usage

### Constructing Keycloak instance

It supports the original 3 flavours of constructing a `KeycloakInstance`.

```'dart'
// This will find the keycloak.json file in the root path
final keycloak = KeycloakInstance();

// This will load the config file at the given path
final keycloak = KeycloakInstance('other_keycloak.json');

// This will construct with the given map
final keycloak = KeycloakInstance.parameters({
    "realm": "demo",
    "authServerUrl": "http://localhost:8080/auth",
    "clientId": "client"
});
```

### Initializing Keycloak

All [flows](https://www.keycloak.org/docs/latest/securing_apps/index.html#flows) are supported. For example, to initialize a Keycloak instance with implicit flow and login immediately:

```'dart'
try {
    final authenticated = await keycloak.init(KeycloakInitOptions(
        flow: 'implicit',
        onLoad: 'login-required'));
    if (authenticated) {
        _loadPage();
    }
} on catch (e) {
    _handleError(e);
}
```

There is one **restriction**: `KeycloakInitOptions.promiseType` must be `'native'` or leave blank. In order to have Dart's `Future` works for all API.

### All Promise based API is Dart's `Future`

As demonstrated above, all `KeycloakInstance` promise based API are converted to Dart's `Future`. You can use the `Future.then()` too if you want:

```'dart'
keycloak.updateToken(55).then((success) {
    if (success) {
        print("Token Refreshed!");
    } else {
        print("Token hasn't expired!");
    }
}).catchError((e) {
    _handleError(e);
});
```

### Registers callbacks

There are a few callback function one can listen to, simply assign a function to such setter. Example:

```'dart'
keycloak.onAuthSuccess = () => print('on auth success');
```

### Handling Exceptions

Not all exception thrown by keycloak JS adapter is a valid `KeycloakError`. It does threw empty exception for some methods. In Dart, it will became a `NullThrownError`.

We can first try to cast the error to `KeycloakError` and see if we can display meaningful error message. If it doesn't, we just display the generic error message. Nonetheless, there is an exception thrown and we should handle the situation.

```'dart'
try {
    final profile = await keycloak.loadUserProfile();
    _displayProfile(profile);
} catch (e) {
    final errorMessage = e is KeycloakError
      ? e.error
      : e?.toString() ?? 'Unknown Error';
    print(errorMessage);
}
```

## Development

### Generating the interop file

1. Must have [js_facade_gen](https://github.com/dart-lang/js_facade_gen?) installed.
2. Go into `/bin` folder and execute `generate_js_interop.sh`:
   - This script assumed you have [keycloak repository](https://github.com/keycloak/keycloak) cloned locally alongside this project's folder.
3. There will be syntax error after the generation (the generator is old and lack of maintenance). You have to repair the generated `/lib/src/js_interop/keycloak.dart`

Here are the repair instructions:

1. It mistakenly generate syntax error at ln 316:

   ```'dart'
   abstract class KeycloakInstance<TPromise extends dynamic /*'native'|dynamic*/, undefined> {
   ```

   To fix it, simply remove the second generic parameter `undefined`, it should be just:

   ```'dart'
   abstract class KeycloakInstance<TPromise extends dynamic /*'native'|dynamic*/> {
   ```

2. It mistakenly thought there was a Keycloak namespace in the keycloak.js, but there isn't. It result in calling all JS functions with a Keycloak namespace in front e.g. `Keycloak.Keycloak()`. To fix it, replace the first line:

   ```'dart'
   @JS("Keycloak")
   ```

   with empty namespace:

   ```'dart'
   @JS()
   ```

3. It mistakenly made all the JavaScript callback functions into a Dart function. To fix them, convert all callback functions to a set function.

   All these functions need to be replace, starting from ln 421: (commented out the replaced generated codes)

   ```'dart
   //external void onReady([bool authenticated]);
   external set onReady(dynamic func);

   //external void onAuthSuccess();
   external set onAuthSuccess(dynamic func);

   //external void onAuthError(KeycloakError errorData);
   external set onAuthError(dynamic func);

   //external void onAuthRefreshSuccess();
   external set onAuthRefreshSuccess(dynamic func);

   //external void onAuthRefreshError();
   external set onAuthRefreshError(dynamic func);

   //external void onAuthLogout();
   external set onAuthLogout(dynamic func);

   //external void onTokenExpired();
   external set onTokenExpired(dynamic func);
   ```

4. `KeycloakTokenParsed` is using outdated types. This is not a code generator issue but an outdated TypeScript definition [issue](https://issues.jboss.org/browse/KEYCLOAK-10745). [PR](https://github.com/keycloak/keycloak/pull/6132) to fix it is active here. This repair is no longer needed once the fix it in.

   Replace the type of `realm_access` and `resources_access` to the responding class respectively: (commented out the replaced generated codes)

   ln 278

   ```'dart'
   // external dynamic /*{ roles: string[] }*/ get realm_access;
   // external set realm_access(dynamic /*{ roles: string[] }*/ v);
   // external List<String> get resource_access;
   // external set resource_access(List<String> v);
   external KeycloakRoles get realm_access;
   external set realm_access(KeycloakRoles v);
   external KeycloakResourceAccess get resource_access;
   external set resource_access(KeycloakResourceAccess v);
   ```

   ln 288

   ```'dart'
   //dynamic /*{ roles: string[] }*/ realm_access,
   //List<String> resource_access});
   KeycloakRoles realm_access,
   KeycloakResourceAccess resource_access});
   ```

### Running the example

Example are a web page demonstrating most of the functionality you can do with this Keycloak Adapter.

#### Setup the Keycloak server

- A local Keycloak server running at `http://localhost:8080`.
- A 'demo' realm setup.
- A 'test_alpha' client.

You can run with your own configuration, just make sure you replace the `keycloak.json` in the `example/` folder.

#### Serve the example

Run it with `webdev serve example:2700`. and visit `http://localhost:2700`.

#### The take-away

This example shows:

- Constructing `KeycloakInstance`
- All 3 [flows](https://www.keycloak.org/docs/latest/securing_apps/index.html#flows) initializations.
- 'login-required' initialization.
- Acquiring user profile, realm access and client access.
- Update token.
- Future APIs.
- Register callbacks.

## Testing

### Setup

All tests assumes:

- A local Keycloak server running at `http://localhost:8080`.
- A 'demo' realm setup.
- 2 clients:

  1. test_alpha
  2. test_beta

### Run

Just run command `pub run build_runner test`.


CHANGELOG.md
## 0.0.1

- A complete functioning Dart wrapper for Keycloak.js.

## 0.0.0

- Initial version, created by Stagehand


test/authentication_test.dart
@TestOn("browser")
import 'package:test/test.dart';

import 'package:keycloak/keycloak.dart';

void main() {
  group('Initialization.', () {
    test('Create keycloak instance by config file', () async {
      final keycloak = KeycloakInstance('keycloak.json');
      await keycloak.init(KeycloakInitOptions());

      expect(keycloak.realm, 'demo');
      expect(keycloak.clientId, 'test_alpha');
    });

    test('Create keycloak instance by parameters', () async {
      final keycloak = KeycloakInstance.parameters({
        "realm": "demo",
        "authServerUrl": "http://localhost:8080/auth",
        "clientId": "test_beta"
      });
      await keycloak.init(KeycloakInitOptions());

      expect(keycloak.realm, 'demo');
      expect(keycloak.clientId, 'test_beta');
    });

    test('Create 2 separated instances', () async {
      final keycloakAlpha = KeycloakInstance();
      final keycloakBeta = KeycloakInstance.parameters({
        "realm": "demo",
        "authServerUrl": "http://localhost:8080/auth",
        "clientId": "test_beta"
      });

      await Future.wait([
        keycloakAlpha.init(KeycloakInitOptions()),
        keycloakBeta.init(KeycloakInitOptions())
      ]);

      expect(keycloakAlpha.realm, 'demo');
      expect(keycloakBeta.realm, 'demo');
      expect(keycloakAlpha.clientId, 'test_alpha');
      expect(keycloakBeta.clientId, 'test_beta');
    });
  });
}


example/main.dart
import 'dart:html';

import 'package:keycloak/keycloak.dart';

/// A global flag to determine should we use 'login-required' when initializing Keycloak
bool loginRequired = true;

/// Start Up will always run after the URL has has been parsed, which determine which `flow` to use.
///
/// Here, we can see how to register callbacks function with `KeycloakInstance`.
/// And how to initialize the instance as well.
///
/// Base on the result of the `init()`, it then decide to show different UI for aunthenticard user vs none.
void _startUp([String flow]) async {
  final keycloak = KeycloakInstance();

  // Register callback and print out in the console
  keycloak.onReady = ([bool authenticate]) => print('on ready $authenticate');
  keycloak.onAuthSuccess = () => print('on auth success');
  keycloak.onAuthError =
      ([KeycloakError error]) => print('on auth error: ${error.error}');
  keycloak.onAuthRefreshSuccess = () => print('on auth refresh success');
  keycloak.onAuthRefreshError = () => print('on auth refresh error');
  keycloak.onAuthLogout = () => print('on logout');
  keycloak.onTokenExpired = () => print('on token expired');

  // Init Keycloak
  bool autenticated = false;
  try {
    autenticated = await keycloak.init(KeycloakInitOptions(
        flow: flow, onLoad: loginRequired ? 'login-required' : ''));
  } catch (e) {
    _errorPage(e, 'Initializing Keycloak');
    return;
  }

  // Checking `keycloak.authenticated` achieved the same thing here.
  if (autenticated) {
    _userSection(keycloak);
  } else {
    _loginSection(keycloak);
  }
}

/// Login Section will be shown if the user is not authenticated.
///
/// It then will show only the Login button which call `Keycloak.login()`.
/// At the same time, it demonstrate that an unauthenticated instance still carries
/// useful informations.
void _loginSection(KeycloakInstance keycloak) {
  final loginButton = querySelector('#button1') as ButtonElement;
  (querySelector('#button2') as ButtonElement).hidden = true;
  (querySelector('#button3') as ButtonElement).hidden = true;

  // Login Button
  loginButton.text = 'Login';
  loginButton.onClick.listen((event) async {
    await keycloak.login();
  });

  final currentSituation = '''
    <h3>${keycloak.flow} Flow: Login</h3>
    <strong>Server:</strong> ${keycloak.authServerUrl} <br>
    <strong>Realm:</strong> ${keycloak.realm} <br>
    <strong>Client:</strong> ${keycloak.clientId} <br>
    ''';

  querySelector('#output').innerHtml = currentSituation;
}

/// User Section will be shown if the user is aunthenticated.
///
/// There are 3 possible actions here:
/// 1. User can click on profile button, then we made an asynchonous call to `Keycloak.loadUserProfile()`,
///    And display the user information when the `Future` succeeded.
///
/// 2. User can click on update button (if the flow is not `imlicit`).
///    It will too make an asynchonous call to `Keycloak.updateToken()` and return later with result.
///
/// 3. User can logout.
///
void _userSection(KeycloakInstance keycloak) {
  final profileButton = querySelector('#button1') as ButtonElement;
  final updateButton = querySelector('#button2') as ButtonElement;
  final logoutButton = querySelector('#button3') as ButtonElement;

  // Show Profile Button
  profileButton.text = 'Show Profile';
  profileButton.onClick.listen((event) async {
    try {
      final profile = await keycloak.loadUserProfile();
      String profileText = '''
      <h4>${profile.username}'s Profile</h4>
      <strong>Name:</strong> ${profile.firstName} ${profile.lastName}<br>
      <strong>Email:</strong> ${profile.email}<br>
      <strong>Created on:</strong> ${profile.createdTimestamp}<br>
    ''';

      querySelector('#profile').innerHtml = profileText;
    } catch (e) {
      _errorPage(e, 'Load User Profile');
      return;
    }
  });

  // Update Token Button
  updateButton.text = 'Update Token';

  // Do not allow updateToken for implicit flow as refreshToken is not available
  if (keycloak.flow == 'implicit') {
    updateButton.disabled = true;
  }
  updateButton.onClick.listen((event) {
    // Showing another usage of `Future`.
    keycloak.updateToken(55).then((success) {
      if (success) {
        _displayAuthenticatedInfo(
            keycloak, "${keycloak.flow} Flow: Token Updated!");
      } else {
        _displayAuthenticatedInfo(
            keycloak, "${keycloak.flow} Flow: Token hasn't expired!");
      }
    }).catchError((e) {
      _errorPage(e, 'Update Token');
    });
  });

  // Logout Button
  logoutButton.text = 'Logout';
  logoutButton.onClick.listen((event) async {
    await keycloak.logout();
  });

  _displayAuthenticatedInfo(keycloak, "${keycloak.flow} Flow: Authenticated!");
}

/// Display Authenticated Info will print out information for aunthenticated user.
///
/// These include the tokens and also roles.
///
void _displayAuthenticatedInfo(KeycloakInstance keycloak, String header) {
  final currentSituation = '''
    <h3>$header</h3>
    <strong>idToken:</strong> ${_ellipsi(keycloak.idToken)} <br>
    <strong>token:</strong> ${_ellipsi(keycloak.token)} <br>
    <strong>refreshToken:</strong> ${_ellipsi(keycloak.refreshToken)} <br>
    <strong>realm roles:</strong> ${keycloak.realmAccess.roles} <br>
    <strong>resource access:</strong> ${keycloak.resourceAccess[keycloak.clientId]?.roles} <br>
    <h4>From Parsed Token</h4>
    <strong>session state:</strong> ${keycloak.tokenParsed.session_state} <br>
    <strong>realm roles:</strong> ${keycloak.tokenParsed.realm_access.roles} <br>
    <strong>resources roles:</strong> ${keycloak.tokenParsed.resource_access[keycloak.clientId]?.roles} <br>
    ''';

  querySelector('#output').innerHtml = currentSituation;
}

void main() async {
  _handleLoginRequiredCheck();

  // Parse the URL hash at the beginning and whenever user click on a link.
  _handleNavigation(window.location.hash);
  window.onHashChange.listen((event) {
    _handleNavigation(window.location.hash);
  });
}

void _handleLoginRequiredCheck() {
  // Acquire the saved decision about using 'login-required'
  loginRequired =
      window.localStorage['login-required'] == 'true' ? true : false ?? false;

  final checkBox = querySelector('#login-required') as CheckboxInputElement;
  checkBox.checked = loginRequired;
  checkBox.onClick.listen((event) {
    loginRequired = checkBox.checked;
    window.localStorage['login-required'] = loginRequired.toString();
  });
}

void _handleNavigation(String hash) {
  if (hash.isEmpty) {
    _startUp();
  } else {
    // A special handling of URL hash to make sure we use the correct `KeycloakInitOption` after
    // redirected back from login page.
    var flowName = RegExp(r'^#\w*').stringMatch(hash);
    switch (flowName) {
      case '#standard':
      case '#implicit':
      case '#hybrid':
        _highlightFlow(flowName);
        flowName = flowName.substring(1);
        break;
      default:
        flowName = null;
        break;
    }

    _startUp(flowName);
  }
}

void _highlightFlow(String flowName) {
  final anchors = querySelectorAll('.activeLink');
  for (var anchor in anchors) {
    anchor.className = 'flowLink';
  }

  final activeAnchor = querySelector(flowName) as AnchorElement;
  activeAnchor.className = 'activeLink';
}

void _errorPage(error, [String additionalInfo]) {
  // Not all exception thrown by keycloak JS adapter is a valid `KeycloakError`.
  // It does threw empty exception for some methods.
  // In Dart, it will became a `NullThrownError`
  final errorMessage = error is KeycloakError
      ? error.error
      : error?.toString() ?? 'Unknown Error';
  querySelector('#error').text = 'Error: $errorMessage. $additionalInfo';
}

String _ellipsi(String s) {
  if (s?.isNotEmpty ?? false) {
    return s.replaceRange(0, s.length - 10, '...');
  }
  return s;
}


lib/src/js_interop/promise.dart
@JS()
library js_promise;

import 'dart:async' show Completer;

import 'package:js/js.dart' show allowInterop, JS;

typedef void ResolveFn<T>(T value);
typedef void RejectFn(dynamic error);

@JS("JSON.parse")
external dynamic parse(obj);

@JS('Promise')
class Promise<T> extends Thenable<T> {
  external Promise(void callback(ResolveFn<T> resolveFn, RejectFn rejectFn));
  external static Promise<List> all(List<Promise> values);
  external static Promise reject(dynamic error);
  external static Promise resolve(dynamic value);
}

@JS('Thenable')
abstract class Thenable<T> {
  // ignore: non_constant_identifier_names
  external Thenable JS$catch([RejectFn rejectFn]);
  external Thenable then([ResolveFn<T> resolveFn, RejectFn rejectFn]);
}

Future<T> promiseToFuture<J, T>(Promise<J> promise,
    [T unwrapValue(J jsValue)]) {
  // handle if promise object is already a future.
  Completer<T> completer = new Completer();
  promise.then(allowInterop((value) {
    T unwrapped;
    if (unwrapValue == null) {
      unwrapped = value as T;
    } else if (value != null) {
      unwrapped = unwrapValue(value);
    }
    completer.complete(unwrapped);
  }), allowInterop((error) {
    completer.completeError(error);
  }));
  return completer.future;
}


lib/src/js_interop/jsHelper.dart
@JS()
library js_helper;

import 'dart:async' show Completer;

import 'package:js/js.dart' show allowInterop, JS;

typedef void ResolveFn<T>(T value);
typedef void RejectFn(dynamic error);

@JS("JSON.parse")
external dynamic parse(obj);

@JS('Promise')
class Promise<T> extends Thenable<T> {
  external Promise(void callback(ResolveFn<T> resolveFn, RejectFn rejectFn));
  external static Promise<List> all(List<Promise> values);
  external static Promise reject(dynamic error);
  external static Promise resolve(dynamic value);
}

@JS('Thenable')
abstract class Thenable<T> {
  // ignore: non_constant_identifier_names
  external Thenable JS$catch([RejectFn rejectFn]);
  external Thenable then([ResolveFn<T> resolveFn, RejectFn rejectFn]);
}

Future<T> promiseToFuture<J, T>(Promise<J> promise,
    [T unwrapValue(J jsValue)]) {
  // handle if promise object is already a future.
  Completer<T> completer = new Completer();
  promise.then(allowInterop((value) {
    T unwrapped;
    if (unwrapValue == null) {
      unwrapped = value as T;
    } else if (value != null) {
      unwrapped = unwrapValue(value);
    }
    completer.complete(unwrapped);
  }), allowInterop((error) {
    completer.completeError(error);
  }));
  return completer.future;
}


lib/src/keycloak_instance.dart
import 'dart:convert' show json;

import 'package:js/js.dart' show allowInterop;
import 'package:js/js_util.dart' show getProperty;

import 'js_interop/keycloak.dart' as js;
import 'js_interop/jsHelper.dart';

/// For callback function registration
typedef void Func();

/// A wrapper for `KeyclockInstance` from the interop.
/// It provides a more Dart friendly interface to replace bare JS interop methods.
///
/// It supports the original 3 flavours of constructing a `KeycloakInstance`.
///
/// ```'dart'
/// // This will find the keycloak.json file in the root path
/// final keycloak = KeycloakInstance();
///
/// // This will load the config file at the given path
/// final keycloak = KeycloakInstance('other_keycloak.json');
///
/// // This will construct with the given map
/// final keycloak = KeycloakInstance.parameters({
///     "realm": "demo",
///     "authServerUrl": "http://localhost:8080/auth",
///     "clientId": "client"
/// });
/// ```
///
/// All [flows](https://www.keycloak.org/docs/latest/securing_apps/index.html#flows)
/// are supported. For example, to initialize a Keycloak instance with
/// implicit flow and login immediately:
///
/// ```'dart'
/// try {
///     final authenticated = await keycloak.init(KeycloakInitOptions(
///         flow: 'implicit',
///         onLoad: 'login-required'));
///     if (authenticated) {
///         _loadPage();
///     }
/// } on KeycloakError catch (e) {
///     print('error $e');
///     return;
/// }
/// ```
/// There is one **restriction**: `KeycloakInitOptions.promiseType` must be
/// `'native'` or leave blank. In order to have Dart's `Future` works for all API.
///
/// All promise based API are converted to Dart's `Future`.
/// You can use the `Future.then()` too if you want:
///
/// ```'dart'
/// keycloak.updateToken(55).then((success) {
///     if (success) {
///         print("Token Refreshed!");
///     } else {
///         print("Token hasn't expired!");
///     }
/// }).catchError((e) {
///     if (e is KeycloakError) {
///         _errorPage(e);
///     }
/// });
/// ```
///
/// There are a few callback function one can listen to,
/// simply assign a function to such setter. Example:
///
/// ```'dart'
/// keycloak.onAuthSuccess = () => print('on auth success');
/// ```
///
/// For the rest of the getter/setter, it just forward the calls.
///
class KeycloakInstance {
  /// The real `KeycloakInstance` JavaScript Object
  js.KeycloakInstance<Promise> _kc;

  /// Is true if the user is authenticated, false otherwise.
  bool get authenticated => _kc.authenticated;
  set authenticated(bool v) => _kc.authenticated = v;

  /// The user id.
  String get subject => _kc.subject;
  set subject(String v) => _kc.subject = v;

  /// Response mode passed in init (default value is `'fragment'`).
  String get responseMode => _kc.responseMode;
  set responseMode(String v) => _kc.responseMode = v;

  /// Response type sent to Keycloak with login requests. This is determined
  /// based on the flow value used during initialization, but can be overridden
  /// by setting this value.
  String get responseType => _kc.responseType;
  set responseType(String v) => _kc.responseType = v;

  /// Flow passed in init.
  String get flow => _kc.flow;
  set flow(String v) => _kc.flow = v;

  /// The realm roles associated with the token.
  js.KeycloakRoles get realmAccess => _kc.realmAccess;
  set realmAccess(js.KeycloakRoles v) => _kc.realmAccess = v;

  /// The resource roles associated with the token.
  KeycloakResourceAccess get resourceAccess =>
      KeycloakResourceAccess(_kc.resourceAccess);
  set resourceAccess(KeycloakResourceAccess v) =>
      _kc.resourceAccess = v.jsObject;

  /// The base64 encoded token that can be sent in the Authorization header in
  /// requests to services.
  String get token => _kc.token;
  set token(String v) => _kc.token = v;

  /// The parsed token as a JavaScript object.
  KeycloakTokenParsed get tokenParsed => KeycloakTokenParsed(_kc.tokenParsed);
  set tokenParsed(KeycloakTokenParsed v) => _kc.tokenParsed = v.jsObject;

  /// The base64 encoded refresh token that can be used to retrieve a new token.
  String get refreshToken => _kc.refreshToken;
  set refreshToken(String v) => _kc.refreshToken = v;

  /// The parsed refresh token as a JavaScript object.
  js.KeycloakTokenParsed get refreshTokenParsed => _kc.refreshTokenParsed;
  set refreshTokenParsed(js.KeycloakTokenParsed v) =>
      _kc.refreshTokenParsed = v;

  /// The base64 encoded ID token.
  String get idToken => _kc.idToken;
  set idToken(String v) => _kc.idToken = v;

  /// The parsed id token as a JavaScript object.
  js.KeycloakTokenParsed get idTokenParsed => _kc.idTokenParsed;
  set idTokenParsed(js.KeycloakTokenParsed v) => _kc.idTokenParsed = v;

  /// The estimated time difference between the browser time and the Keycloak
  /// server in seconds. This value is just an estimation, but is accurate
  /// enough when determining if a token is expired or not.
  num get timeSkew => _kc.timeSkew;
  set timeSkew(num v) => _kc.timeSkew = v;

  bool get loginRequired => _kc.loginRequired;
  set loginRequired(bool v) => _kc.loginRequired = v;

  String get authServerUrl => _kc.authServerUrl;
  set authServerUrl(String v) => _kc.authServerUrl = v;

  String get realm => _kc.realm;
  set realm(String v) => _kc.realm = v;

  String get clientId => _kc.clientId;
  set clientId(String v) => _kc.clientId = v;

  String get clientSecret => _kc.clientSecret;
  set clientSecret(String v) => _kc.clientSecret = v;

  String get redirectUri => _kc.redirectUri;
  set redirectUri(String v) => _kc.redirectUri = v;

  String get sessionId => _kc.sessionId;
  set sessionId(String v) => _kc.sessionId = v;

  js.KeycloakProfile get profile => _kc.profile;
  set profile(js.KeycloakProfile v) => _kc.profile = v;

  get userInfo => _kc.userInfo;
  set userInfo(v) => _kc.userInfo = v;

  /// Called when the adapter is initialized.
  set onReady(Func f) => _kc.onReady = allowInterop(f);

  /// Called when a user is successfully authenticated.
  set onAuthSuccess(Func f) => _kc.onAuthSuccess = allowInterop(f);

  /// Called if there was an error during authentication.
  set onAuthError(Func f) => _kc.onAuthError = allowInterop(f);

  /// Called when the token is refreshed.
  set onAuthRefreshSuccess(Func f) =>
      _kc.onAuthRefreshSuccess = allowInterop(f);

  /// Called if there was an error while trying to refresh the token.
  set onAuthRefreshError(Func f) => _kc.onAuthRefreshError = allowInterop(f);

  /// Called if the user is logged out (will only be called if the session
  /// status iframe is enabled, or in Cordova mode).
  set onAuthLogout(Func f) => _kc.onAuthLogout = allowInterop(f);

  /// Called when the access token is expired. If a refresh token is available
  /// the token can be refreshed with Keycloak#updateToken, or in cases where
  /// it's not (ie. with implicit flow) you can redirect to login screen to
  /// obtain a new access token.
  set onTokenExpired(Func f) => _kc.onTokenExpired = allowInterop(f);

  /// Constructing an instance with an optional file path
  ///
  /// If `configFilePath` wasn't defined, It will search for 'keycloak.json' in the root.
  KeycloakInstance([configFilePath]) {
    _kc = js.Keycloak(configFilePath);
  }

  /// Constructing an instance with key value pair.
  /// It is akin to JavaScript Object definition but in fact is a Dart `Map`.
  /// `params` will be convert from a `map` to a `Object`.
  KeycloakInstance.parameters(Map params) {
    _kc = js.Keycloak(parse(json.encode(params)));
  }

  /// Called to initialize the adapter.
  Future<bool> init(js.KeycloakInitOptions options) async {
    // If user define a `promiseType` and it is not 'native'
    if (options.promiseType != null && options.promiseType != 'native') {
      throw ArgumentError.value(options.promiseType,
          "We only support 'native' promiseType, given ${options.promiseType}");
    }

    // Forced native Promise in order to use our `Promise` interop
    options.promiseType = 'native';

    return promiseToFuture(_kc.init(options));
  }

  /// Redirects to login form.
  Future login([js.KeycloakLoginOptions options]) =>
      promiseToFuture(_kc.login(options));

  /// Redirects to logout.
  Future logout([dynamic options]) => promiseToFuture(_kc.logout(options));

  /// Redirects to registration form.
  /// set to `'register'`.
  Future register([dynamic options]) => promiseToFuture(_kc.register(options));

  /// Redirects to the Account Management Console.
  Future accountManagement() => promiseToFuture(_kc.accountManagement());

  /// Returns the URL to login form.
  String createLoginUrl([js.KeycloakLoginOptions options]) =>
      _kc.createLoginUrl(options);

  /// Returns the URL to logout the user.
  String createLogoutUrl([dynamic options]) => _kc.createLogoutUrl(options);

  /// Returns the URL to registration page.
  /// `action` is set to `'register'`.
  String createRegisterUrl([js.KeycloakLoginOptions options]) =>
      _kc.createRegisterUrl(options);

  /// Returns the URL to the Account Management Console.
  String createAccountUrl() => _kc.createAccountUrl();

  /// Returns true if the token has less than `minValidity` seconds left before
  /// it expires.
  bool isTokenExpired([num minValidity]) => _kc.isTokenExpired(minValidity);

  /// If the token expires within `minValidity` seconds, the token is refreshed.
  /// If the session status iframe is enabled, the session status is also
  /// checked.
  /// still valid, or if the token is no longer valid.
  /// @example
  /// ```js
  /// keycloak.updateToken(5).success(function(refreshed) {
  /// if (refreshed) {
  /// alert('Token was successfully refreshed');
  /// } else {
  /// alert('Token is still valid');
  /// }
  /// }).error(function() {
  /// alert('Failed to refresh the token, or the session has expired');
  /// });
  Future<bool> updateToken(num minValidity) =>
      promiseToFuture(_kc.updateToken(minValidity));

  /// Clears authentication state, including tokens. This can be useful if
  /// the application has detected the session was expired, for example if
  /// updating token fails. Invoking this results in Keycloak#onAuthLogout
  /// callback listener being invoked.
  void clearToken() => _kc.clearToken();

  /// Returns true if the token has the given realm role.
  bool hasRealmRole(String role) => _kc.hasRealmRole(role);

  /// Returns true if the token has the given role for the resource.
  bool hasResourceRole(String role, [String resource]) =>
      _kc.hasResourceRole(role, resource);

  /// Loads the user's profile.
  Future<js.KeycloakProfile> loadUserProfile() =>
      promiseToFuture(_kc.loadUserProfile());

  /// @private Undocumented.
  Future loadUserInfo() => promiseToFuture(_kc.loadUserInfo());
}

/// A wrapper for `KeycloakResourceAccess` from the interop.
///
/// Index signature is not yet supported by JavaScript interop.
/// So, to keep using the same API of `KeycloakResrouceAccess['client']`,
/// we have to wrap it with a dart class and use `getProperty`
///
class KeycloakResourceAccess {
  final js.KeycloakResourceAccess jsObject;

  KeycloakResourceAccess(this.jsObject);

  js.KeycloakRoles operator [](String name) => getProperty(jsObject, name);
}

/// A wrapper for `KeycloakTokenParsed` from the interop.
///
/// It involved returning a `KeycloakResourceAccess`, which require a special handling.
///
class KeycloakTokenParsed {
  final js.KeycloakTokenParsed jsObject;

  num get exp => jsObject.exp;
  set exp(num v) => jsObject.exp = v;
  num get iat => jsObject.iat;
  set iat(num v) => jsObject.iat = v;
  String get nonce => jsObject.nonce;
  set nonce(String v) => jsObject.nonce = v;
  String get sub => jsObject.sub;
  set sub(String v) => jsObject.sub = v;
  String get session_state => jsObject.session_state;
  set session_state(String v) => jsObject.session_state = v;
  js.KeycloakRoles get realm_access => jsObject.realm_access;
  set realm_access(js.KeycloakRoles v) => jsObject.realm_access = v;
  KeycloakResourceAccess get resource_access =>
      KeycloakResourceAccess(jsObject.resource_access);
  set resource_access(KeycloakResourceAccess v) =>
      jsObject.resource_access = v.jsObject;

  KeycloakTokenParsed(this.jsObject);
}
